src/main/java/seedu/task/logic/parser/AddParser.java

```diff
@@ -42,50 +37,42 @@ public AddParser() {}
-                    return new AddTaskCommand(
-                        taskMatcher.group("name").trim(),
-                        taskMatcher.group("description").trim(),
-                        dl.get().trim()
-                    );
-                } else {
-                    return new AddTaskCommand(
-                            taskMatcher.group("name").trim(),
-                            taskMatcher.group("description").trim()
-                    );
+        if (args.isEmpty()) {
+            return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
+        }
+        
+        ArgumentTokenizer argsTokenizer = new ArgumentTokenizer(descriptionPrefix, deadlinePrefix, durationPrefix);
+        argsTokenizer.tokenize(args);
+        
+        try {           
+            String name = argsTokenizer.getPreamble().get();
+            Optional <String> description = argsTokenizer.getValue(descriptionPrefix);
+            Optional <String> duration = argsTokenizer.getValue(durationPrefix);
+            Optional <String> deadline = argsTokenizer.getValue(deadlinePrefix);
+            
+            if (duration.isPresent()) { //Only events have duration
+                try {
+                    return new AddEventCommand(name, description.orElse(""), duration.orElse(""));
+                } catch (IllegalValueException ive) {
+                    return new IncorrectCommand(ive.getMessage());
+                } catch (NoSuchElementException nsee) {
+                    return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
+                }
+            } else {
+                try {
+                    return new AddTaskCommand(name, description.orElse(""), deadline.orElse(""));             
+                } catch (IllegalValueException ive) {
+                    return new IncorrectCommand(ive.getMessage());
+                } catch (NoSuchElementException nsee) {
+                    return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
                 }
-                
-            } catch (IllegalValueException ive) {
-                return new IncorrectCommand(ive.getMessage());
-            }
-        } else if (eventMatcher.matches()){
-            try {
-                return new AddEventCommand(
-                        eventMatcher.group("name").trim(),
-                        eventMatcher.group("description").trim(),
-                        eventMatcher.group("duration").trim()
-                );
-            } catch (IllegalValueException ive) {
-                return new IncorrectCommand(ive.getMessage());
             }
-        }else {
+        } catch (NoSuchElementException nsee) {
             return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
+        } catch (EmptyValueException e) {
```

Comment:
is this catch for empty arguments? if that's the case, you can delete your earlier code at line 41-43.